<img src="images/bannerugentdwengo.png" alt="Banner" width="400"/>

<div>
    <font color=#690027 markdown="1">
        <h1>NUMERIEK BEPALEN VAN DE LIMIET VAN EEN RIJ</h1>
    </font>
</div>

<div class="alert alert-box alert-success">
In de wiskundeles leer je limieten van rijen en functies berekenen a.d.h.v. bepaalde technieken en formules. Maar je kan een limiet ook numeriek benaderen.<br>
In deze notebook zal je de limiet van een rij bepalen door de limiet numeriek te benaderen.
</div>

<div class="alert alert-box alert-success">
Om de limiet van een rij numeriek te benaderen is het belangrijk dat je een goed beeld krijgt van de termen met een grote index. Eigenlijk moet je zien te achterhalen naar waar de termen van de rij evolueren als je de index naar oneindig laat gaan.
</div>

<div>
    <font color=#690027 markdown="1">
        <h2>1. De harmonische rij</h2> 
    </font>
</div>

Beschouw de harmonische rij $u$ met expliciet voorschrift $u_{n} = \frac{1}{n}$, met $n \geq 1$.<br>
Deze rij is $1, \frac{1}{2}, \frac{1}{3}, \frac{1}{4}, ...$

Om enkele termen van deze rij door Python te kunnen laten zien, heb je een expliciet voorschrift nodig van deze rij en een manier om enkele termen te laten berekenen a.d.h.v. dat voorschrift. 

<div>
    <font color=#690027 markdown="1">
        <h3>1.1 De rij definiëren en enkele termen laten zien</h3> 
    </font>
</div>

Merk op dat $u_{0}=\frac{1}{1}$ en dat elke term dus gezien kan worden als een breuk met een $1$ in de teller. Bij de term met index $0$ staat er een $1$ in de noemer, de term met index $1$ heeft een $2$ in de noemer, bij index $2$ staat er in de noemer een $3$, enz. 

Definieer a.d.h.v. de volgende code-cel de **algemene term** $u_{n}$ van deze rij in Python.

Laat nu de eerste 10 termen van de rij berekenen en laat ze zien.

Combineer het voorgaande om een functie te definiëren waarmee je een bepaald aantal termen van de harmonische rij kuan laten berekenen. Laat vervolgens tien termen zien. 

#### Tien termen van de harmonische rij

#### Oefening 1.1
- Is de harmonische rij stijgend, dalend of schommelend?
- Vermoed je dat de rij convergent of divergent is? 

Antwoord:

<div>
    <font color=#690027 markdown="1">
        <h2>2. Op zoek naar de limiet van de harmonische rij</h2> 
    </font>
</div>

<div>
    <font color=#690027 markdown="1">
        <h3>2.1  Honderd termen</h3> 
    </font>
</div>

Uit tien termen kan je nog niet opmaken wat de limiet van de harmonische rij is.<br>
Bepaal meer termen. 

Hieruit kan je nog steeds niet opmaken welke de limiet van de harmonische rij is.<br>
Met zoveel termen als uitvoer wordt het ook snel onoverzichtelijk.

<div>
    <font color=#690027 markdown="1">
        <h3>2.2  Tot de duizendste term in de rij</h3> 
    </font>
</div>

#### Voorbeeld 
- Toon enkele termen, maar niet de opeenvolgende; neem grotere stappen. Zorg ervoor dat je tot de duizendste term in de rij komt, maar zonder ze allemaal te laten zien. 

De volgende code-cel biedt een mogelijke oplossing voor deze opdracht.

In [94]:
# termen berekenen 
u = harmonische_rij(1000)

# enkele van eerste 1000 termen van rij tonen in stappen van 50 (behalve eerste stap, die is 49) via f-string 
print(f"u{0} = {u[0]}") 
for n in range(49, 1000, 50):
    print(f"u{n} = {u[n]}") 

u0 = 1.0
u49 = 0.02
u99 = 0.01
u149 = 0.006666666666666667
u199 = 0.005
u249 = 0.004
u299 = 0.0033333333333333335
u349 = 0.002857142857142857
u399 = 0.0025
u449 = 0.0022222222222222222
u499 = 0.002
u549 = 0.0018181818181818182
u599 = 0.0016666666666666668
u649 = 0.0015384615384615385
u699 = 0.0014285714285714286
u749 = 0.0013333333333333333
u799 = 0.00125
u849 = 0.001176470588235294
u899 = 0.0011111111111111111
u949 = 0.0010526315789473684
u999 = 0.001


In de vorige code-cel werden de eerste 1000 termen allemaal berekend, maar ze werden niet allemaal getoond. Er werd dus veel overbodig rekenwerk gedaan. <br>
In wat volgt, ga je **efficienter** te werk.

Je definieert eerst een functie die enkel de termen van de harmonische rij zal berekenen, die je ook wilt laten zien. 

In [95]:
# aantal termen van harmonische rij berekenen tot bepaalde index, maar met indexsprongen
def harmonische_rij_indexsprong(startindex, eindindex, sprong):
    """Termen van harmonische rij berekenen tot bepaalde index met indexsprong."""
    u = []
    for n in range(startindex, eindindex+1, sprong):   
        term = 1/(n+1)
        u.append(term)
    return u

Met de volgende code-cel laat je de termen van de harmonische rij genereren en zien vanaf index 49 t.e.m. 999.

In [96]:
# harmonische rij van index 49 tot en met index 999 met indexsprongen van 50
# index 49, 99, 149, ...
u_sprong = harmonische_rij_indexsprong(49, 999, 50)
print(u_sprong)

[0.02, 0.01, 0.006666666666666667, 0.005, 0.004, 0.0033333333333333335, 0.002857142857142857, 0.0025, 0.0022222222222222222, 0.002, 0.0018181818181818182, 0.0016666666666666668, 0.0015384615384615385, 0.0014285714285714286, 0.0013333333333333333, 0.00125, 0.001176470588235294, 0.0011111111111111111, 0.0010526315789473684, 0.001]


De volgende code-cel geeft een mooiere uitvoer; de termen worden getoond met hun index, dus met hun positie in de rij, erbij.

In [97]:
# berekende termen met index tonen via f-string en enumerate()
for i, term in enumerate(u_sprong):
    print(f"u_sprong{(i+1)*50-1} = {term}") 

u_sprong49 = 0.02
u_sprong99 = 0.01
u_sprong149 = 0.006666666666666667
u_sprong199 = 0.005
u_sprong249 = 0.004
u_sprong299 = 0.0033333333333333335
u_sprong349 = 0.002857142857142857
u_sprong399 = 0.0025
u_sprong449 = 0.0022222222222222222
u_sprong499 = 0.002
u_sprong549 = 0.0018181818181818182
u_sprong599 = 0.0016666666666666668
u_sprong649 = 0.0015384615384615385
u_sprong699 = 0.0014285714285714286
u_sprong749 = 0.0013333333333333333
u_sprong799 = 0.00125
u_sprong849 = 0.001176470588235294
u_sprong899 = 0.0011111111111111111
u_sprong949 = 0.0010526315789473684
u_sprong999 = 0.001


Dit is echter nog steeds onvoldoende om de limiet te benaderen. Gaan tot een index van 999 is niet ver genoeg in de rij. Anderzijds heb je nu misschien wel al een beter beeld wat de limiet van deze rij is. 

<div>
    <font color=#690027 markdown="1">
        <h3>2.3  Tot de miljoenste term in de rij</h3> 
    </font>
</div>

In [107]:
# harmonische rij van index 4999 tot en met index 999999 met indexsprongen van 15000
u_sprong_verder = harmonische_rij_indexsprong(4999, 999999, 15000)
print(u_sprong_verder)

[0.0002, 5e-05, 2.857142857142857e-05, 2e-05, 1.5384615384615384e-05, 1.25e-05, 1.0526315789473684e-05, 9.090909090909091e-06, 8e-06, 7.142857142857143e-06, 6.451612903225806e-06, 5.882352941176471e-06, 5.405405405405405e-06, 5e-06, 4.651162790697674e-06, 4.347826086956521e-06, 4.081632653061224e-06, 3.846153846153846e-06, 3.636363636363636e-06, 3.4482758620689654e-06, 3.2786885245901638e-06, 3.125e-06, 2.9850746268656716e-06, 2.8571428571428573e-06, 2.7397260273972604e-06, 2.631578947368421e-06, 2.531645569620253e-06, 2.4390243902439023e-06, 2.352941176470588e-06, 2.2727272727272728e-06, 2.197802197802198e-06, 2.1276595744680853e-06, 2.061855670103093e-06, 2e-06, 1.941747572815534e-06, 1.8867924528301887e-06, 1.834862385321101e-06, 1.7857142857142857e-06, 1.7391304347826088e-06, 1.6949152542372882e-06, 1.652892561983471e-06, 1.6129032258064516e-06, 1.5748031496062992e-06, 1.5384615384615385e-06, 1.5037593984962406e-06, 1.4705882352941177e-06, 1.4388489208633094e-06, 1.4084507042253521

In [108]:
# enkele van eerste 1000000 termen van rij tonen in stappen van 15000 (behalve eerste stap, die is 4999)
print(f"u_sprong_verder{4999} = {u_sprong_verder[0]}") 
for i, term in enumerate(u_sprong_verder[1:]):
    print(f"u_sprong_verder{(i+2)*10000-1} = {term}") 

u_sprong_verder4999 = 0.0002
u_sprong_verder19999 = 5e-05
u_sprong_verder29999 = 2.857142857142857e-05
u_sprong_verder39999 = 2e-05
u_sprong_verder49999 = 1.5384615384615384e-05
u_sprong_verder59999 = 1.25e-05
u_sprong_verder69999 = 1.0526315789473684e-05
u_sprong_verder79999 = 9.090909090909091e-06
u_sprong_verder89999 = 8e-06
u_sprong_verder99999 = 7.142857142857143e-06
u_sprong_verder109999 = 6.451612903225806e-06
u_sprong_verder119999 = 5.882352941176471e-06
u_sprong_verder129999 = 5.405405405405405e-06
u_sprong_verder139999 = 5e-06
u_sprong_verder149999 = 4.651162790697674e-06
u_sprong_verder159999 = 4.347826086956521e-06
u_sprong_verder169999 = 4.081632653061224e-06
u_sprong_verder179999 = 3.846153846153846e-06
u_sprong_verder189999 = 3.636363636363636e-06
u_sprong_verder199999 = 3.4482758620689654e-06
u_sprong_verder209999 = 3.2786885245901638e-06
u_sprong_verder219999 = 3.125e-06
u_sprong_verder229999 = 2.9850746268656716e-06
u_sprong_verder239999 = 2.8571428571428573e-06
u_spr

Er zijn opnieuw veel termen als uitvoer, waardoor het onoverzichtelijk is. 

#### Opdracht 2.2
Zoek hiervoor een oplossing! Hoe zou je snel kunnen vorderen in de indices van de termen van de rij, maar zonder al te veel termen als uitvoer te hebben? En met een duidelijk zicht op hoe de termen evolueren naarmate de index groter wordt.

<div>
    <font color=#690027 markdown="1">
        <h3>2.3 De limiet van de rij</h3> 
    </font>
</div>

Je merkt dat de termen steeds kleiner worden, maar wel steeds positief blijven. De rij nadert naar nul.
$$ \lim_{n \to +\infty} u_{n} = 0$$

<div>
    <font color=#690027 markdown="1">
        <h2>3. Oefeningen</h2> 
    </font>
</div>

#### Oefening 3.1

Bepaal de limiet van de rij $u_{n} = -n^{2}$, met $n \geq 0$.

#### Oefening 3.2

Bepaal de limiet van de rij $u_{n} = 5 n - 20 \sin(n) $, met $n \geq 0$.

<div>
    <font color=#690027 markdown="1">
        <h2>4. Grafische voorstelling</h2> 
    </font>
</div>

Je zou de termen van de rij ook grafisch kunnen voorstellen. Dan krijg je vast sneller een zicht op de situatie en mogelijke waarde van de limiet. Voor de werkwijze kan je terecht bij de notebooks over limieten van functies.  

<img src="images/cclic.png" alt="Banner" align="left" width="100"/><br><br>
Notebook Python voor Numerieke methodes, zie <a href="http://www.aiopschool.be">AI Op School</a>, van F. wyffels & N. Gesquière, is in licentie gegeven volgens een <a href="http://creativecommons.org/licenses/by-nc-sa/4.0/">Creative Commons Naamsvermelding-NietCommercieel-GelijkDelen 4.0 Internationaal-licentie</a>. 